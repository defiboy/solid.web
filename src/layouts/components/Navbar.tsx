import React from 'react'

import { Action, ActionCreator } from 'redux'

import { Button, Dropdown, Icon, Menu } from 'antd'
import styled, { css } from 'styled-components'

const ButtonGroup = Button.Group

const Wrapper = styled.header`
  grid-area: header;
  background-color: #272727;
  display: flex;
  color: white;
  align-items: center;
  padding-left: 1em;
  padding-right: 1em;
  display: grid;
  grid-template-columns: 16em auto 16em;
  grid-template-areas: 'leftbutton title rightbutton';
`

const ButtonRightArea = styled.div`
  grid-column-start: 1;
  grid-column-end: 3;
  grid-area: leftbutton;
  display: flex;
  justify-content: flex-start;
`

const ButtonGroupItem = styled(ButtonGroup)`
  padding-right: 1em;
`

const SmallButton = styled(Button)`
  height: 26px !important;
  padding: 0 6px !important;
`

const styles = css`
  font-weight: 100;
  font-size: 1.2em;
  color: white;
  font-family: 'Helvetica'
  grid-area: title;
  grid-column-start: 2;
  grid-column-end: 3;
  justify-self: center;
`

const NavbarTitle = styled(({ children, reverse, palette, theme, ...props }) =>
  React.createElement(`h1`, props, children)
)`
  ${styles}
`

interface Props {
  onNewConnectionClick: ActionCreator<Action>
  onNewContractDefinitionClick: ActionCreator<Action>
}

export const Navbar: React.FC<Props> = ({ onNewConnectionClick, onNewContractDefinitionClick }: Props) => {
  const onDropDownClickk = ({ key }: any) => {
    if (key === 'contract') {
      onNewContractDefinitionClick()
    } else if (key === 'connection') {
      onNewConnectionClick()
    }
  }

  const renderMenu = () => {
    return (
      <Menu onClick={onDropDownClickk}>
        <Menu.Item data-testid="navbar-menu-contract" key="contract">Contract Definition</Menu.Item>
        <Menu.Item data-testid="navbar-menu-connection" key="connection">Connection</Menu.Item>
      </Menu>
    )
  }

  return (
    <Wrapper>
      <ButtonRightArea>
        <ButtonGroupItem>
          <SmallButton data-testid={`button-navbar-new`} type="primary" size="small" onClick={onNewConnectionClick}>
            <Icon type="plus" />
            New
          </SmallButton>
          <Dropdown overlay={renderMenu}>
            <SmallButton data-testid={`navbar-menu`} type="primary" size="small">
              <Icon type="down" />
            </SmallButton>
          </Dropdown>
        </ButtonGroupItem>
        {/* <ButtonGroupItem>
          <SmallButton type="primary" size="small">
            <Icon type="fall" />
            Profiler
          </SmallButton>
        </ButtonGroupItem> */}
      </ButtonRightArea>
      <NavbarTitle>Solid Studio 1.0</NavbarTitle>
    </Wrapper>
  )
}
