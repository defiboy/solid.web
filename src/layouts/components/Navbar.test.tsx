
import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';

import { Navbar } from './Navbar'

describe('Navbar', () => {

  const onNewConnectionClickMock = jest.fn()
  const onNewContractInstanceClickMock = jest.fn()

  beforeEach(() => {
    onNewConnectionClickMock.mockClear()
    onNewContractInstanceClickMock.mockClear()
  })

  test('navbar ui items are rendered', () => {
    const { getByTestId, getByText } = render(
      
      <Navbar onNewConnectionClick={onNewConnectionClickMock}
        onNewContractInstanceClick={onNewContractInstanceClickMock}/>)

    const newButton = getByTestId('button-navbar-new')
    const title = getByText(/solid/i)

    expect(newButton).toHaveAttribute('type', 'button')
    expect(newButton).toBeInTheDocument()
    expect(title).toBeInTheDocument()
  })

  test('when new connection button is clicked, onNewConnectionClick is called', () => {
    const { getByTestId } = render(
      
      <Navbar onNewConnectionClick={onNewConnectionClickMock}
        onNewContractInstanceClick={onNewContractInstanceClickMock}/>)

    const newButton = getByTestId('button-navbar-new')
    fireEvent.click(newButton)

    expect(onNewConnectionClickMock).toHaveBeenCalled()
    expect(onNewConnectionClickMock).toHaveBeenCalledTimes(1)
  })


  test('when new connection from dropdown is clicked, onNewConnectionClick is called', async () => {
    const { getByTestId } = render(
      
      <Navbar onNewConnectionClick={onNewConnectionClickMock}
        onNewContractInstanceClick={onNewContractInstanceClickMock}/>)

    const dropdownButton = getByTestId('navbar-menu')
    fireEvent.mouseOver(dropdownButton)

    await wait(() => {
      expect(getByTestId('navbar-menu-contract')).toBeInTheDOM()
      expect(getByTestId('navbar-menu-connection')).toBeInTheDOM()
    })

    const connectionOptionInDropdown = getByTestId('navbar-menu-connection')
    fireEvent.click(connectionOptionInDropdown)

    expect(onNewConnectionClickMock).toHaveBeenCalled()
    expect(onNewConnectionClickMock).toHaveBeenCalledTimes(1)
  })


  test('when new contract instance button is clicked, onNewContractInstanceClick is called', async () => {
    const { getByTestId } = render(
      
      <Navbar onNewConnectionClick={onNewConnectionClickMock}
        onNewContractInstanceClick={onNewContractInstanceClickMock}/>)

    const dropdownButton = getByTestId('navbar-menu')
    fireEvent.mouseOver(dropdownButton)

    await wait(() => {
      expect(getByTestId('navbar-menu-contract')).toBeInTheDOM()
      expect(getByTestId('navbar-menu-connection')).toBeInTheDOM()
    })

    const contractOptionInDropdown = getByTestId('navbar-menu-contract')
    fireEvent.click(contractOptionInDropdown)

    expect(onNewContractInstanceClickMock).toHaveBeenCalled()
    expect(onNewContractInstanceClickMock).toHaveBeenCalledTimes(1)
  })

  test('renders navbar snapshot', async () => {

    const { getByTestId, baseElement } = render(
      
      <Navbar onNewConnectionClick={onNewConnectionClickMock}
        onNewContractInstanceClick={onNewContractInstanceClickMock}/>)

    const dropdownButton = getByTestId('navbar-menu')
    fireEvent.mouseOver(dropdownButton)

    await wait(() => {
      expect(getByTestId('navbar-menu-contract')).toBeInTheDOM()
      expect(getByTestId('navbar-menu-connection')).toBeInTheDOM()
    })

    expect(baseElement.innerHTML).toMatchSnapshot()
  })
})
