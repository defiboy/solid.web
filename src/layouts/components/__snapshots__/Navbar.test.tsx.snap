// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders navbar snapshot 1`] = `
"<style>
      [ant-click-animating-without-extra-node='true']::after, .ant-click-animating-node {
        --antd-wave-shadow-color: buttonface;
      }</style><div><header class=\\"sc-bdVaJa bUbuxh\\"><div class=\\"sc-bwzfXH hIAJaZ\\"><div class=\\"ant-btn-group sc-htpNat jVmkYh\\"><button data-testid=\\"button-navbar-new\\" type=\\"button\\" class=\\"ant-btn sc-bxivhb eqTrKp ant-btn-primary ant-btn-sm\\"><i aria-label=\\"icon: plus\\" class=\\"anticon anticon-plus\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"plus\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z\\"></path><path d=\\"M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z\\"></path></svg></i><span>New</span></button><button data-testid=\\"navbar-menu\\" type=\\"button\\" class=\\"ant-btn sc-bxivhb eqTrKp ant-dropdown-trigger ant-dropdown-open ant-btn-primary ant-btn-sm\\" style=\\"\\"><i aria-label=\\"icon: down\\" class=\\"anticon anticon-down\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></i></button></div></div><h1 class=\\"sc-ifAKCX iKPNWQ\\">Solid Studio 1.0</h1></header></div><div style=\\"position: absolute; top: 0px; left: 0px; width: 100%;\\"><div><div class=\\"ant-dropdown ant-dropdown-placement-bottomLeft\\" style=\\"left: -999px; top: -969px;\\"><ul class=\\"ant-dropdown-menu ant-dropdown-menu-light ant-dropdown-menu-root ant-dropdown-menu-vertical\\" role=\\"menu\\" tabindex=\\"0\\"><li data-testid=\\"navbar-menu-contract\\" class=\\"ant-dropdown-menu-item\\" role=\\"menuitem\\">Contract Definition</li><li data-testid=\\"navbar-menu-connection\\" class=\\"ant-dropdown-menu-item\\" role=\\"menuitem\\">Connection</li></ul></div></div></div>"
`;
